[MM-29137] - post inline editing

michelengelen opened the pull request:
#### Summary
~~This PR is in draft phase and will be completed after the productivity sprint is over!~~
~~Just wanted to have this in for visibility!~~

#### related PR
adding a feature flag to the server repo
https://github.com/mattermost/mattermost-server/pull/19210

#### Ticket Link
[MM-29137](https://mattermost.atlassian.net/browse/MM-29137)

#### Screenshots
TBD

#### Release Note
TBD
```release-note
allows for inline editing of posts
```

matthewbirtch reviewed (changes requested):
Thanks @michelengelen this is a great start. I'm going to leave any UI polish items out of this review for now, but I noticed some functional things on first review that I thought I would call out:

1. When I edit a post using the `Edit` option from the post actions menu, we should set the focus on the inline field immediately. Right now if I click edit and start typing, it focuses on the main message input. Because of this, pressing `ENTER` or `ESC` doesn't do anything until I have clicked in to the input. 
2. We will need to update the style and highlighting to make the active field more obvious too, but I'm going to assume that will come later.
2. When using the `UP ARROW` shortcut to edit the last post, we should set the focus on the inline field
3. Can we add the instruction below the inline edit input "ENTER to Save, ESC to Cancel" (this message would also adapt depending on user setting for ENTER to send)
4. I assume attachment editing/adding be in a separate PR?
5. When I click the reaction button within the inline field, the emoji picker is hidden
6. I wonder if we should disable the post hover actions for the message that is currently being edited
![image](https://user-images.githubusercontent.com/2040554/133815589-03f4e34f-a309-4d0f-8109-aae8aa1cee3d.png)
7. With CRT Off, when attempting to edit a reply in the RHS Thread view, it showed the inline edit input in the main channel view. 
8. With CRT turned on and I try to edit a post from within the RHS or the Global Threads view, editing doesn't work at all.

michelengelen commented:
@esethna yes, I will certainly do so ... its just that there is a lot of stuff on the table atm! So please be patient ... it will proceed! 🚋

hmhealey commented:
Removing my request for review for the moment, but you can re-add me when you have a chance to work on this again.

michelengelen commented:
hey @matthewbirtch and @esethna ... I think this is ready for a new round of reviews.

Regarding your points @matthewbirtch:

>     1. When I edit a post using the `Edit` option from the post actions menu, we should set the focus on the inline field immediately. Right now if I click edit and start typing, it focuses on the main message input. Because of this, pressing `ENTER` or `ESC` doesn't do anything until I have clicked in to the input.

done

>     2. We will need to update the style and highlighting to make the active field more obvious too, but I'm going to assume that will come later.

this can be done rather quick now, since there is only one class being added to the wrapping container that handles this kind of stuff.

>     3. When using the `UP ARROW` shortcut to edit the last post, we should set the focus on the inline field

done

>     4. Can we add the instruction below the inline edit input "ENTER to Save, ESC to Cancel" (this message would also adapt depending on user setting for ENTER to send)

done

>     5. I assume attachment editing/adding be in a separate PR?

yes

>     6. When I click the reaction button within the inline field, the emoji picker is hidden

done ... it now overflows to the top (as it should be)

>     7. I wonder if we should disable the post hover actions for the message that is currently being edited

absolutely ... AAAAAND ... done 😄 

>     8. With CRT Off, when attempting to edit a reply in the RHS Thread view, it showed the inline edit input in the main channel view.

it is now possible to edit in the RHS and it focusses the post correctly

>     9. With CRT turned on and I try to edit a post from within the RHS or the Global Threads view, editing doesn't work at all.

same as 8.

Let me know if you have any input on styling ... and, of course, when you find any bugs.

I am going to remove the draft status as well.

Thanks!!! 🙇🏼‍♂️

esethna commented:
This is sweet! I found a few things in a brief test, will let Matt go more into the UI:

1. Styling issues with various post types if the edited messages is not the last in the channel (ie CRT off replies, consecutive posts etc): 
![image](https://user-images.githubusercontent.com/14333569/142912387-d7fc709f-938c-48bf-85b7-b86432c7873a.png)
![image](https://user-images.githubusercontent.com/14333569/142912568-18b1d49d-ce98-4dad-a641-186330de941d.png)


2. The click to reply for threads is activating in the inline editing input area causing conflicts of which text area is focused. Please reach out to Kyriakos if you need some more context on that feature

matthewbirtch reviewed (changes requested):
In addition to what @esethna mentioned, I'm seeing the following issues:

4. If I use the `UP` arrow key to edit my last point, the input doesn't seem to get the right background color.

![image](https://user-images.githubusercontent.com/2040554/142919700-a2616551-79a4-4de7-9592-601553fc7543.png)

5. The area where the help text `ENTER to Save, ESC to Cancel is getting highlighted or it just has the wrong background color in some cases. 
![image](https://user-images.githubusercontent.com/2040554/142919860-1f1c9325-52cc-45e0-b91d-1a0c2c7e65cb.png)

6. If I edit a message, the main input field for the channel does not get focused on again. We don't want to lose this function, so we should make sure it gets focused again - both in the center main channel area, global threads view and the RHS
7. If I have `⌘+ENTER` set as my send shortcut (Settings > Advanced > Send Messages on ⌘+ENTER) that is not reflected in the help text. We should update the help text to reflect this setting
8. I tried editing a long post and it scrolls the post I'm editing very low in the viewport. See the GIF example. 

![2021-11-22 14 15 29](https://user-images.githubusercontent.com/2040554/142921812-26cb74e5-bb1e-46c8-86fb-2efa24e90829.gif)

9. Similar to the above, there seems to be an awkward shift between default and editing state in the RHS
<img src="https://user-images.githubusercontent.com/2040554/142925071-44062e46-27e0-4083-87ed-26f08e50d88c.gif" width="300">


10. Editing a long post in the RHS is awkward because it wraps to a very narrow width. It look like there is 90px of padding on the right side - can we reduce this to allow for a wider text width?
![image](https://user-images.githubusercontent.com/2040554/142923131-d21760d5-fe6b-4162-94
b4-abf8a272816a.png)
11. Wondering if we can try a smooth(quick) transition to the editing state - I have a feeling it would be difficult to make it performant, but what do you think? ([see prototype example](https://www.figma.com/proto/fVLSLjGls2d2LUvCqk2JMS/%E2%9C%85MM-29137-Message-Editing?page-id=0%3A1&node-id=524%3A17015&viewport=465%2C48%2C0.06&scaling=min-zoom&starting-point-node-id=524%3A17015&show-proto-sidebar=1)

michelengelen commented:
Hey @esethna ... I fixed quite a bit here, so feel free to take a look again:

>     1. Styling issues with various post types if the edited messages is not the last in the channel (ie CRT off replies, consecutive posts etc):

should be fixed now

>     2. The click to reply for threads is activating in the inline editing input area causing conflicts of which text area is focused. Please reach out to Kyriakos if you need some more context on that feature

clicking the textinput does not trigger opening the thread view in RHS anymore, when the post is in "being edited" state

>     3. I can get stuck in a state where the hover effect doesn't disappear after opening the in-line editor and closing it

this is fixed now... great find BTW! :D

Same goes for @matthewbirtch:

>     4. If I use the `UP` arrow key to edit my last point, the input doesn't seem to get the right background color.

should be fixed now

>     5. The area where the help text `ENTER to Save, ESC to Cancel is getting highlighted or it just has the wrong background color in some cases.

should be fixed as well (same problem as 1.)

>     6. If I edit a message, the main input field for the channel does not get focused on again. We don't want to lose this function, so we should make sure it gets focused again - both in the center main channel area, global threads view and the RHS

fixed

>     7. If I have `⌘+ENTER` set as my send shortcut (Settings > Advanced > Send Messages on ⌘+ENTER) that is not reflected in the help text. We should update the help text to reflect this setting

should now show the correct key

>     8. I tried editing a long post and it scrolls the post I'm editing very low in the viewport. See the GIF example.

the edit-field now scrolls into view. There is only one thing I would like to mention here: the way it is currently built it will **always** scroll the textarea to the bottom of the parent container. even if it is fully visible. We could implement some kind of check, but wanted to touch base with you first on this. 

>     9. Similar to the above, there seems to be an awkward shift between default and editing state in the RHS

the only way to lessen the "awkwardness" is to add a transition, but that might be OOS for this PR ... but I am 0/5 on this ... agreed?

>     10. Editing a long post in the RHS is awkward because it wraps to a very narrow width. It look like there is 90px of padding on the right side - can we reduce this to allow for a wider text width?

fixed

>     11. Wondering if we can try a smooth(quick) transition to the editing state - I have a feeling it would be difficult to make it performant, but what do you think? ([see prototype example](https://www.figma.com/proto/fVLSLjGls2d2LUvCqk2JMS/%E2%9C%85MM-29137-Message-Editing?page-id=0%3A1&node-id=524%3A17015&viewport=465%2C48%2C0.06&scaling=min-zoom&starting-point-node-id=524%3A17015&show-proto-sidebar=1)

see point 9

esethna reviewed (approved):
Absolutely amazing, really great work @michelengelen @matthewbirtch!

matthewbirtch commented:
Thanks for resolving those @michelengelen, great work. I found a few new issues:

12. In the RHS - the reactions icon button is displaying outside of the container:

<img width="504" alt="Screen Shot 2021-11-23 at 1 55 40 PM" src="https://user-images.githubusercontent.com/2040554/143087274-ae188072-2e9e-4874-b788-01bec132b74d.png">

13. Also, I'm seeing the hover state staying on with the last post I edited in the center channel:

![image](https://user-images.githubusercontent.com/2040554/143088057-e5562dc7-df1a-4778-939a-3a09cf9c1391.png)

14. In mobile webview, hovering over the post in edit mode changes the background to white. Not a bit deal though since the majority of users in this viewport size will be on mobile devices where hover doesn't exist. There are some on desktop who shrink their browser though.

15. I tried using the emoji picker while in edit mode and it gets clipped to the container and I can't see the whole picker. NOTE: this doesn't seem to be happening in the RHS - only in the center channel.
![image](https://user-images.githubusercontent.com/2040554/143088988-cc872610-d042-41db-896f-1138510ee43c.png)

And responding to a few comments above:

> > ```
> > 8. I tried editing a long post and it scrolls the post I'm editing very low in the viewport. See the GIF example.
> > ```
> 
> 
> the edit-field now scrolls into view. There is only one thing I would like to mention here: the way it is currently built it will **always** scroll the textarea to the bottom of the parent container. even if it is fully visible. We could implement some kind of check, but wanted to touch base with you first on this.

Maybe we can have a quick chat about this. I don't fully understand. The goal is to have the whole text input visible when in edit mode. Can we not scroll the channel area so that it fits?

> > ```
> > 9. Similar to the above, there seems to be an awkward shift between default and editing state in the RHS
> > ```
> 
> 
> the only way to lessen the "awkwardness" is to add a transition, but that might be OOS for this PR ... but I am 0/5 on this ... Thoughts?

Maybe we make the height of the text input the same height as max height of the truncated post? (Where the 'show more' button displays). Then the transition from a long, truncated post to edit mode wouldn't be as jarring. Just an idea - might cause other issues though.

michelengelen commented:
@matthewbirtch comments to your findings below.

> Thanks for resolving those @michelengelen, great work. I found a few new issues:
> 
>     12. In the RHS - the reactions icon button is displaying outside of the container:
> 

fixed

> 
>     13. Also, I'm seeing the hover state staying on with the last post I edited in the center channel:
> 

fixed

> 
>     14. In mobile webview, hovering over the post in edit mode changes the background to white. Not a bit deal though since the majority of users in this viewport size will be on mobile devices where hover doesn't exist. There are some on desktop who shrink their browser though.
>

fixed ... now does not have any hover effect anymore since it is already marked as "active/focussed".

> 
>     15. I tried using the emoji picker while in edit mode and it gets clipped to the container and I can't see the whole picker. NOTE: this doesn't seem to be happening in the RHS - only in the center channel.
>

fixed

> 
> And responding to a few comments above:
> 
> > > ```
> > > 8. I tried editing a long post and it scrolls the post I'm editing very low in the viewport. See the GIF example.
> > > ```
> > 
> > 
> > the edit-field now scrolls into view. There is only one thing I would like to mention here: the way it is currently built it will **always** scroll the textarea to the bottom of the parent container. even if it is fully visible. We could implement some kind of check, but wanted to touch base with you first on this.
> 
> Maybe we can have a quick chat about this. I don't fully understand. The goal is to have the whole text input visible when in edit mode. Can we not scroll the channel area so that it fits?

Sure ... do you mind scheduling something this week?

> > > ```
> > > 9. Similar to the above, there seems to be an awkward shift between default and editing state in the RHS
> > > ```
> > 
> > 
> > the only way to lessen the "awkwardness" is to add a transition, but that might be OOS for this PR ... but I am 0/5 on this ... Thoughts?
> 
> Maybe we make the height of the text input the same height as max height of the truncated post? (Where the 'show more' button displays). Then the transition from a long, truncated post to edit mode wouldn't be as jarring. Just an idea - might cause other issues though.

We can talk about that in the meeting as well if you don't mind.

matthewbirtch commented:
> > ```
> > 13. Also, I'm seeing the hover state staying on with the last post I edited in the center channel:
> > ```

Thanks @michelengelen looks like the majority of the identified issues are resolved, but I'm still seeing this one

![image](https://user-images.githubusercontent.com/2040554/143911761-9052177b-bccf-4afb-b635-454f3440dcbd.png)

Also, I booked some time for us to discuss the remaining issues tomorrow. Thanks!

larkox commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,481 @@
+
+    useEffect(() => textboxRef?.current?.focus(), []);
+    useEffect(() => {
+        const parentId = `${editingPost.isRHS ? 'rhsP' : 'p'}ost_${editingPost.postId}`;
```
I would do `...'rhsPost' : 'post'}_${...`. We are only saving a few characters, and this way is more readable.

larkox commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,481 @@
+        return () => {
+            document.removeEventListener('paste', handlePaste);
+        };
+    });
```
Do we want to run this on every render?

larkox replied:
Probably not worth it, but have you consider make the handlePaste a callback, and make the effect only about adding and removing the event (and depending just on the handler)?

michelengelen replied:
nice catch ... I forgot to attach the `[]` ... thanks.

Unfortunately you have to define these inside the `useEffect` to be effective ... :/

larkox commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,481 @@
+const TOP_OFFSET = 0;
+const RIGHT_OFFSET = 10;
+
+const EditPost = ({editingPost, actions, ...rest}: Props): JSX.Element => {
```
Really personal preference, so 0/5, but I would prefer the rest of the props to be called `props` instead of `rest`.

michelengelen replied:
I just followed the "convention" from the code-base here: we only use `...props` once but `...rest` several times ... I would have preferred `props` as well

larkox commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,481 @@
+            return;
+        }
+
+        actions.addMessageIntoHistory(updatedPost.message);
```
Is this desirable? The edited text to become part of the history of messages?

michelengelen replied:
good question  .... I am not 100% sure what the history is used for, so maybe not, since the original message is already in there

larkox replied:
This history is for showing different inputs when you hit `ctrl+up`. It is really useful when you are typing commands.

michelengelen replied:
well ... first of all: thanks for clarifying that. ... but I cannot use that hotkey on my machine (on localhost and community), since it does nothing. Is it still in place?

larkox commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,481 @@
+            Date.now(),
+            0,
+            caretPosition,
+        ) as {allowSending: boolean; ignoreKeyPress?: boolean};
```
Why is the `as` needed? Can we make it so it is not?

michelengelen replied:
it is not ...

larkox reviewed (approved):
Nothing really blocking from my side, just a bunch of minor things here and there.

michelengelen commented on actions/views/textbox.js:
```diff
@@ -22,10 +22,3 @@ export function setShowPreviewOnEditChannelHeaderModal(showPreview) {
-
-export function setShowPreviewOnEditPostModal(showPreview) {
-    return {
-        type: ActionTypes.SET_SHOW_PREVIEW_ON_EDIT_POST_MODAL,
```
did i not remove it already?

michelengelen replied:
no, i did not ... will remove it in the next commit.

michelengelen commented on components/edit_post/index.ts:
```diff
@@ -2,72 +2,64 @@
         channelId,
-        shouldShowPreview: showPreviewOnEditPostModal(state),
-        maxPostSize: parseInt(config.MaxPostSize, 10) || Constants.DEFAULT_CHARACTER_LIMIT,
+        maxPostSize: config.MaxPostSize ? parseInt(config.MaxPostSize, 10) : Constants.DEFAULT_CHARACTER_LIMIT,
```
nice catch

michelengelen commented on actions/post_actions.test.js:
```diff
@@ -302,13 +302,6 @@ describe('Actions.Posts', () => {
         expect(testStore.getActions()).toEqual([]);
     });
 
-    test('hideEditPostModal', async () => {
```
maybe only in conjunction with `setEditingPost`, but I am 1/5 that this is very helpful as well

michelengelen commented on components/post_view/post_info/post_info.tsx:
```diff
@@ -184,17 +192,15 @@ export default class PostInfo extends React.PureComponent<Props, State> {
-    getDotMenu = (): HTMLDivElement => {
-        return this.dotMenuRef.current as HTMLDivElement;
-    };
+    getDotMenu = (): HTMLDivElement => this.dotMenuRef.current as HTMLDivElement;
```
it is indeed not really needed, but the components implementing it had it typed out to return a `HTMLDivElement` from this function, so it got casted on the `current` (which in itself is a smell). I changed the definitions to also accept `null`

michelengelen commented on components/post_view/post_info/post_info.tsx:
```diff
@@ -320,17 +326,21 @@ export default class PostInfo extends React.PureComponent<Props, State> {
 
-    componentDidUpdate(prevProps: Props) {
-        const {shortcutReactToLastPostEmittedFrom, isLastPost} = this.props;
+    componentDidUpdate(prevProps: Props): void {
```
there is a linter rule in place, but only as warning

michelengelen commented on components/post_view/post_info/post_info.tsx:
```diff
@@ -320,17 +326,21 @@ export default class PostInfo extends React.PureComponent<Props, State> {
         }
+
+        if (!prevProps.isPostBeingEdited && isPostBeingEdited) {
+            this.props.handleDropdownOpened(false);
```
realtively simple answer ... we need to update the dropdownState to being closed ... otherwise it would stay open after we finish the post editing ... we could rename that to be more descriptive, but IMO this adds another layer of abstraction that is not really needed

michelengelen commented on components/rhs_comment/rhs_comment.jsx:
```diff
@@ -146,6 +152,10 @@ export default class RhsComment extends React.PureComponent {
         }
+
+        if (!prevProps.isPostBeingEdited && isPostBeingEdited) {
+            this.handleDropdownOpened(false);
```
same answer as before :D

michelengelen commented on components/rhs_comment/rhs_comment.jsx:
```diff
@@ -434,16 +448,12 @@ export default class RhsComment extends React.PureComponent {
-        let postClass = '';
 
         if (post.failed) {
-            postClass += ' post-failed';
```
it was not ... I searched for it everywhere, but could not find any usages in the code-base

michelengelen commented on selectors/posts.js:
```diff
@@ -11,6 +11,15 @@ import {getGlobalItem} from 'selectors/storage';
+export function getIsPostBeingEditedInRHS(state, postId) {
+    const editingPost = getEditingPost(state);
+
+    return editingPost.isRHS && state.views.posts.editingPost.postId === postId;
```
good question ... I must have forgotten to replace that ... will change that now

michelengelen commented on components/rhs_comment/__snapshots__/rhs_comment.test.jsx.snap:
```diff
@@ -111,36 +111,32 @@ exports[`components/RhsComment should match snapshot 1`] = `
         </div>
```
added

jgilliam17 commented:
E2Es running [here](https://automation-dashboard.vercel.app/cycles/3de39445-79b9-44a2-8214-0005dc1409f2)

matthewbirtch commented:
> > > ```
> > > 13. Also, I'm seeing the hover state staying on with the last post I edited in the center channel:
> > > ```

@michelengelen this looks to be resolved now

jgilliam17 commented:
@michelengelen here's the [E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/94086-16806c9-ee-ent-webapp-pr-8882-mm-ee-test:16806c9/mochawesome.html)
After a quick look, there are at least 23 failing due to missing `editPostModal`

matthewbirtch reviewed (changes requested):
@michelengelen the long posts are working well now, but shorter posts that are higher up on the scroll view seem to jump down to the bottom when I edit. 

![2021-12-04 09 25 24](https://user-images.githubusercontent.com/2040554/144713100-0d92a046-9514-46ab-bee9-45c2280a6cdd.gif)

michelengelen commented:
> @michelengelen the long posts are working well now, but shorter posts that are higher up on the scroll view seem to jump down to the bottom when I edit.
> 

@matthewbirtch thats a problem that is hard to fix without the usage of an additional package, since the built-in method only knows "black-and-white" ... you can either use it and scroll, or don't use it. 

I did not want to add another package just yet, since I am trying to explore if there is a better solution for this. I am having a look on monday! 🚀

jgilliam17 commented:
Thanks @michelengelen 
Looks great, looking forward to using this on community soon.
Few issue 
1. Autocomplete list adds horizontal scroll bar
<img width="671" alt="Screen Shot 2021-12-07 at 3 24 46 PM" src="https://user-images.githubusercontent.com/52937121/145103069-fd46cc45-efcd-488b-9b55-8d77c167b42d.png">

2. When inline edit post is engaged, other post menus will open tucked under it 
<img width="950" alt="Screen Shot 2021-12-07 at 3 31 55 PM" src="https://user-images.githubusercontent.com/52937121/145103347-02eb6af8-c82f-4d99-8bb4-d016bbfe7b95.png">

3. Pinned and saved highlighting is broken 
<img width="552" alt="Screen Shot 2021-12-07 at 3 36 20 PM" src="https://user-images.githubusercontent.com/52937121/145103439-0478cc6d-53fc-4af1-a876-0954bc976a78.png">

4. When selecting "edit" on RHS Pinned posts, Mentions or on Saved posts, inline edit opens on main channel view
5. Inline edit is not dismissed when changing channels. Not sure if this is intended. @matthewbirtch thoughts?
 note:  it is dismissed if you start editing a new post, but there is a chance you may lose edited text on a previous post 
for example:
Start editing a post on Town Square, don't save
Switch to Off Topic then back to Town Square  > inline edit is still open, edited text could be lost if you start to edit a different post

matthewbirtch commented:
> 5\. Inline edit is not dismissed when changing channels. Not sure if this is intended. @matthewbirtch thoughts?

Ya, I think that's reasonable to cancel the editing if you switch channels

michelengelen commented:
@jgilliam17  and @matthewbirtch ... I think this is ready for another round of reviews.

>
>     1. Autocomplete list adds horizontal scroll bar
>

Fixed ... Although I would suggest to rework a part of the `SuggestionList` component that handles the repositioning. IMO the List should be rendered above, when there is not enough space to render below and with the same width and tugged to the left side when it will overflow to the right.
 
> 
>     2. When inline edit post is engaged, other post menus will open tucked under it
> 

fixed

> 
>     3. Pinned and saved highlighting is broken
> 

fixed

>
>     4. When selecting "edit" on RHS Pinned posts, Mentions or on Saved posts, inline edit opens on main channel view
> 

fixed

>
>     5. Inline edit is not dismissed when changing channels. Not sure if this is intended. @matthewbirtch thoughts?
>        note:  it is dismissed if you start editing a new post, but there is a chance you may lose edited text on a previous post
>        for example:
>        Start editing a post on Town Square, don't save
>        Switch to Off Topic then back to Town Square  > inline edit is still open, edited text could be lost if you start to edit a different post
> 

switching channels now removes the editing state

>
>     6. Emoji picker is not working. Nothing can be selected, no hover effect on emoji, simply closes on click without adding emoji to edited post

fixed!

Feel free to reach out when you find new bugs with it! 🙇🏼‍♂️

matthewbirtch commented:
> Fixed ... Although I would suggest to rework a part of the `SuggestionList` component that handles the repositioning. IMO the List should be rendered above, when there is not enough space to render below and with the same width and tugged to the left side when it will overflow to the right.
> 
> >

Completely agree with this. Is it possible to update that in thie PR? Most of the list gets hidden if I'm triggering the autocomplete near the bottom of the screen
![image](https://user-images.githubusercontent.com/2040554/145609468-1eee1d3b-6c88-4e50-9cde-79a1976414b3.png)

Other than that, I think you've resolved everything else!

michelengelen commented:
@matthewbirtch yes, that would be possible

michelengelen commented:
@matthewbirtch my last change should do the trick 🎉

matthewbirtch reviewed (approved):
Great, thanks @michelengelen. Looks great now

esethna commented:
Stoked to get this in! 

cc// @hmhealey and @jgilliam17 for review

jgilliam17 commented:
@michelengelen can you please fix the failing checks, I'd like to run Cypress E2Es again. Thanks 🙂 
( I re-ran the failed checks, but it did nothing)

michelengelen commented:
@jgilliam17 done!

jgilliam17 commented:
Thanks @michelengelen some checks are still failing

hmhealey commented:
I hate to be coming in here with this suggestion after so much work has been done, but shouldn't this be behind a setting or at least a feature flag? It seems like an awfully large change to just have happen suddenly without being able to roll it out across cloud to start

michelengelen commented:
> I hate to be coming in here with this suggestion after so much work has been done, but shouldn't this be behind a setting or at least a feature flag? It seems like an awfully large change to just have happen suddenly without being able to roll it out across cloud to start

You are right and I already thought about this as well... it just did not seem to be that big of a change at first, so that it was not considered. 

I will look into this and come back here when I have a valid solution for it

esethna commented:
@hmhealey @michelengelen this is a really good point. I agree we should introduce this behind a feature flag (I don't think a config is needed)

michelengelen commented:
@hmhealey I added a feature flag to the server repo and snuck everything behind a feature flag getter to enable/disable the stuff for inline editing.

THe flag itself defaults to true, question is how do we enable it on the test-servers? I do not have access to the spiwick configs, so this would need to be done by someone with sufficient acces rights, right?

cc/ @agnivade @cpanato

cpanato commented:
@michelengelen this is something that the cloud SRE can help you, I don't have the access to make those changes

cc @spirosoik @stylianosrigas

stylianosrigas commented:
@michelengelen Please create a Jira ticket and assign it to the Cloud SRE team to help you :)

michelengelen commented:
@jgilliam17 could you please re-run the e2e tests again with the feature enabled? It is: `MM_FEATUREFLAG_INLINEPOSTEDITING=true`
Thanks!!! 🙇🏼

jgilliam17 commented:
@michelengelen E2Es running here https://automation-dashboard.vercel.app/cycles/1f09df32-71fc-4d51-b9cb-961a35cb4d76
Flag enabled `"variables[MM_ENV]=MM_FEATUREFLAGS_INLINEPOSTEDITING=true"`
(I missed previous post before I triggered the run, will re-run if we have to wait for server side PR)

jgilliam17 commented:
There are quite a few failures https://mattermost-cypress-report.s3.amazonaws.com/98955-738f4eb-onprem-ent-webapp-pr-8882-mm-ee-test:738f4eb/mochawesome.html
This seems off, I'll re run after we merge server PR.

michelengelen commented:
@jgilliam17 could you restart the e2e tests please?

jgilliam17 commented:
E2Es running here: https://automation-dashboard.vercel.app/cycles/cfed7505-4adf-42b1-814e-79cc0c74e26f

jgilliam17 commented:
E2E report: https://mattermost-cypress-report.s3.amazonaws.com/100156-0b93288-onprem-ent-webapp-pr-8882-mm-ee-test:0b93288/mochawesome.html

michelengelen commented:
It seems a lot of the failing tests are unrelated to my changes. Could you check as well @saturninoabril ?

I dont think that retention tests have anything to do with this PR, as well as member invites and guest user tests.

I did go through every failing test and tried to find any causes for failing tests that are unrelated to changes in this PR:

## unrelated failing E2E tests

#### error in checkPropTypes.js (```Failed prop type: The prop `teams` is marked as required in `PermissionTeamSchemeSettings`, but its value is `null`.```), which leasd to failing tests:

- **MM-T1550**
- **MM-T2855**
- **MM-T2857**
- **MM-T2862**
- **MM-T2871**
- **MM-T2876**
- **MM-T2878**
- **Team Scheme Guest Permissions Test** (no MM-T* Test number assigned)

#### error in function `checkChannelPermission` (schemes array is empty), leads to failing tests:

- **MM-23018** (2 failing assertions)
- **MM-24379**

#### error in Promise (```TypeError: can't access property Symbol.iterator, `action.data.policies` is null```), which leads to failing tests:

- **MM-T4093**
- **MM-T4097**
- **MM-T4100**

#### misc

it seems these tests are failing due to the image not being loaded when the action/assertion is performed since the file preview button is only displayed when the image is fully loaded

- **MM-T187**
- **MM-T329**
- **MM-T331**
- **MM-T332**
- **MM-T688**
- **MM-T2261**

jgilliam17 commented:
I re-ran the E2Es, report can be viewed here: https://mattermost-cypress-report.s3.amazonaws.com/101044-a4ee2fb-onprem-ent-webapp-pr-8882-mm-ee-test:a4ee2fb/mochawesome.html
for comparison master E2E run: https://mattermost-cypress-report.s3.amazonaws.com/100952-onprem-ent-master-mm-ee-test:test/mochawesome.html
There are total of 4 failed tests on the PR that differ from the master report
MM-T99
MM-T2140
these 2 don't appear to be related to the PR changes
MM-T613
MM-T951

michelengelen commented:
Hi @jgilliam17: I fixed MM-T99 and MM-T2140 with my latest commit. I think with this it is ready for a manual QA-review. Thanks a lot for your patience so far! 🙇🏼‍♂️ 

@saturninoabril any objections?

jgilliam17 commented:
✅ Latest [E2E report](https://mattermost-cypress-report.s3.amazonaws.com/101142-onprem-ent-master-mm-ee-test:test/mochawesome.html) looks good - no PR related failures when compared to [master report ](https://mattermost-cypress-report.s3.amazonaws.com/101265-5f7d41b-onprem-ent-webapp-pr-8882-mm-ee-test:5f7d41b/mochawesome.html)

jgilliam17 commented:
Thanks @michelengelen 
Found few smaller issues
1. Text should be vertically centered 
<img width="405" alt="Screen Shot 2021-12-29 at 10 50 01 AM" src="https://user-images.githubusercontent.com/52937121/147681200-28a9a981-49fe-44e2-b503-2f9278cdf568.png">

2. When editing a reply post on the RHS original post text continues to display below the inline edit textbox
<img width="490" alt="Screen Shot 2021-12-29 at 10 47 34 AM" src="https://user-images.githubusercontent.com/52937121/147681567-f51aa09d-ba0d-4121-b84f-fc670b9800ae.png">

3. We should close inline editing when closing RHS
Currently, closing the RHS without canceling edit, leaves the post in a weird state, highlighted and with missing post menu. 
<img width="1044" alt="Screen Shot 2021-12-29 at 3 24 32 PM" src="https://user-images.githubusercontent.com/52937121/147700711-e98ffb8a-ffd2-4259-8e13-592b26eaef71.png">

michelengelen commented:
@jgilliam17 I fixed the bugs you found. Could you re-review the changes please? Thanks!

jgilliam17 commented:
@michelengelen I am not able to save edited post on Enter. Additionally, no replies are posting on the RHS.
<img width="423" alt="Screen Shot 2022-01-03 at 1 52 15 PM" src="https://user-images.githubusercontent.com/52937121/147968364-4ae2b688-bb8e-43ef-9265-adc7f4a870a4.png">

michelengelen commented:
@jgilliam this is working for me locally ... I will try and re-setup the test instance and see if it is working after that

jgilliam17 commented:
@cpanato would you be able to take a look here? I am not able to save posts on the RHS - server error. It appears that there is an issue with the server configuration, and may not be related to changes on this PR. Thanks 🙂 
<img width="1258" alt="Screen Shot 2022-01-12 at 10 03 39 PM" src="https://user-images.githubusercontent.com/52937121/149258554-16e4966f-2245-4d02-afe7-7c8b7274d4cd.png">

hmhealey commented on packages/mattermost-redux/src/types/config.ts:
```diff
@@ -117,6 +117,7 @@ export type ClientConfig = {
     ExperimentalGroupUnreadChannels: string;
     ExperimentalPrimaryTeam: string;
     ExperimentalTimezone: string;
+    ExperimentalTownSquareIsReadOnly: string;
```
This should be removed again

michelengelen replied:
done

hmhealey commented on utils/a11y_controller.js:
```diff
@@ -817,7 +817,7 @@ export default class A11yController {
 
     handleFocus = (event) => {
-        if (this.lastInputEventIsKeyboard && this.windowIsFocused) {
+        if (this.lastInputEventIsKeyboard && this.windowIsFocused && event.target.id !== 'edit_textbox') {
```
Why do we need a special case for the edit box here?

michelengelen replied:
it placed the `a11y-focussed` class on the element, but there was no easy means of removing that. I do not remember 100%, but it was necessary.

hmhealey replied:
Wouldn't we want to add that class to show focus though? I don't know why this component would behave different than any other textarea with the A11yController active

michelengelen replied:
Ah, I remember now ... it was needed for the case where the user enters edit mode by using the arrow-up key! ... There was no easy way around this, so I added it here.

hmhealey replied:
I still don't understand why this component would have to have special behaviour here though. It looks like this function is only called when the window regains focus, so does that cause the edit box to get stuck with that extra class somehow?


Perhaps we should meet up to chat a bit about this tomorrow since I feel like a few minutes would help me better understand what this is needed for

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+        const parentCR = postListElement?.getBoundingClientRect();
+
+        if (elementCR && parentCR) {
+            const shouldScrollIntoView = shouldScroll(elementCR, parentCR);
```
Sadly not... the behavior of this is very inconsistent. This is the best and smoothest solution for now (without adding another dependency)

michelengelen commented on components/dot_menu/dot_menu.tsx:
```diff
@@ -259,11 +259,12 @@ export class DotMenuClass extends React.PureComponent<Props, State> {
             this.props.location === Locations.CENTER ? 'post_textbox' : 'reply_textbox',
             this.props.post.root_id ? Utils.localizeMessage('rhs_comment.comment', 'Comment') : Utils.localizeMessage('create_post.post', 'Post'),
-            this.props.location === Locations.RHS_ROOT || this.props.location === Locations.RHS_COMMENT,
+            this.props.location === Locations.RHS_ROOT || this.props.location === Locations.RHS_COMMENT || this.props.location === Locations.SEARCH,
```
this was something that got pointed out during PR reviews ... maybe it is just that it is possible to do so at the moment and we don't want to miss that possibility? I am also not 100% sure that this is a good idea, but since it was possible before I added it in.

This would be a good question for @esethna and @matthewbirtch

matthewbirtch replied:
I would be good with that. I actually didn't realize I could edit posts from the search results list, but I would be fine with removing that ability (if we do this we should remove the ability to delete as well) from this context. Users would have to jump to the thread or channel to edit the post. @esethna do you think users use this much?

Alternatively, we could leave the ability to edit from the search results, but direct them to the post in the thread or channel with the post scrolled to and the inline edit input displayed. This sounds complicated though

michelengelen replied:
I am fine with both approaches. Keeping the ability to edit them in the search results just creates more opportunities and less jumping around for the user (maybe the user wants to edit all of his messages where he send a URL or something like that) .. there are definitely use-cases for that :D

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+            const shouldScrollIntoView = shouldScroll(elementCR, parentCR);
+
+            if (shouldScrollIntoView) {
+                const scrollOptions: boolean|ScrollIntoViewOptions = isFirefox() ? {behavior: 'smooth', block: 'nearest'} : false;
```
the method `scrollIntoView` is not very consistent in its behavior. Firefox adds the support for the options, which work way better than just passing the `boolean` for `alignToTop`. The smooth scrolling is way better as well.

For the sake of consistency we could just pass `false` for all browsers, but we would lose a better solution for Firefox. That's something we might need to discuss with UX.

cc/ @matthewbirtch

matthewbirtch replied:
@michelengelen can you clarify what needs to be discussed with UX?

hmhealey replied:
@matthewbirtch When scrolling the post that's being edited into view, we're doing it slightly differently on Firefox. If I understand right, we scroll smoothly to the post instead of jumping to it.

I'd be okay if that's the case then. I usually like consistency, but smooth scrolling is much nicer

michelengelen replied:
@hmhealey is right about this. Firefox supports the smooth scrolling when using `scrollIntoView`. There are other options (aka npm packagers) out there that support smooth scrolling on all browsers, but I did not want to add a new dependency.

We could absolutely discuss this later on if you would like, but for now I would skip this option for the sake of keeping the PR in scope.

matthewbirtch replied:
Agree that smooth scrolling is an improved experience. I'm ok with leaving it in only for Firefox for now, but it would be best to make it consistent for all browsers. I think that would be okay to do in a follow-up PR though

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+        const className = `post-list__dynamic${editingPost.isRHS ? '--RHS' : ''}`;
+        const postListElement = document.getElementsByClassName(className)[0];
+
+        const elementCR = wrapperRef?.current?.getBoundingClientRect();
```
nice catch

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+        textboxRef?.current?.focus();
+    }, []);
+
+    useEffect(() => {
```
absolutely ... i was wondering the same thing, but having it inside the textbox component itself is hardly going to work, since the content needs to live in the upper components. But the idea of adding in a hook to be used when implementing the `CreatePost`/`CreateComment` components is a very good indeed.

I am going to add a comment on that.

michelengelen replied:
BTW @hmhealey: is there already a ticket for porting the textbox component to FC?

hmhealey replied:
Not on its own, but it's something I'm considering that we might do as part of a larger cleanup of the CreatePost/CreateComment components. We're not planning on that any time soon, but maybe late in the quarter if it doesn't interfere with the effort to make it a rich text editor

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+    };
+
+    useEffect(() => {
+        const className = `post-list__dynamic${editingPost.isRHS ? '--RHS' : ''}`;
```
I originally added it because the container was not scrolled into view correctly. I just tested it again using `setTimeout` and now it works. Thanks for catching that ... it makes the whole process a bit better.

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+
+    const handleRefocusAndExit = (refocusId: string|null) => {
+        if (refocusId) {
+            setTimeout(() => {
```
You are right about this ... but only in conjunction with the second comment about resetting the state. I left the setTimeout in because it was refocussing before the unmount happened. I just wanted to make sure that the textbox is only focussed when everything is cleared up, so we do not accidentally land in a weird state.

I removed that and the state resets after checking it is functional locally.

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+            });
+        }
+
+        setCaretPosition(1);
```
see above

hmhealey replied:
I don't think I understand. Is this component not being unmounted when we're done editing then? If that's the case, I'd be worried about having a lot of invisible stuff mounted then

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+    };
+
+    const handleHide = (doRefocus = true) => {
+        setEditText(editingPost.post?.message || '');
```
Nice catch ... I simplified by using `handleRefocusAndExit(editingPost.refocusId || null)` wherever we try to refocus and `handleRefocusAndExit(null)` when showing the `deleteModal`

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+        const data = await actions.editPost(updatedPost as Post);
+
+        if (data) {
+            window.scrollTo(0, 0);
```
I am not sure either .. removing it did not change the behavior, so I removed it for now.

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+            <div>
+                <EmojiPickerOverlay
+                    show={showEmojiPicker}
+                    container={() => textboxRef.current}
+                    target={() => emojiButtonRef.current}
```
done

michelengelen commented on components/edit_post/edit_post.tsx:
```diff
@@ -0,0 +1,494 @@
+                value={editText}
+                channelId={rest.channelId}
+                emojiEnabled={rest.config.EnableEmojiPicker === 'true'}
+                createMessage={Utils.localizeMessage('edit_post.editPost', 'Edit the post...')}
```
done

michelengelen commented on components/edit_post/index.ts:
```diff
@@ -0,0 +1,66 @@
+    const config = getConfig(state);
+    const editingPost = getEditingPost(state);
+    const currentUserId = getCurrentUserId(state);
+    const channelId = editingPost?.post?.channel_id || getCurrentChannelId(state);
```
yes, we do need it at the moment. The problem is that the selector `getEditingPost` cannot safely return a valid `editingPost` object. Otherwise we could just extract `post.channel_id` directly. Although adding a guard should not be too complicated. Let me see what I can do here.

michelengelen replied:
yep ... we can just assume that editingPost is always there when we mount the `EditPost` component, so there is no need for the fallback.

michelengelen commented on components/edit_post/index.ts:
```diff
@@ -0,0 +1,66 @@
+    const deletePermission = isAuthor ? Permissions.DELETE_POST : Permissions.DELETE_OTHERS_POSTS;
+    const editPermission = isAuthor ? Permissions.EDIT_POST : Permissions.EDIT_OTHERS_POSTS;
+
+    const channel = state.entities.channels.channels[channelId] || {};
```
done

michelengelen commented on components/edit_post/index.ts:
```diff
@@ -0,0 +1,66 @@
+    const editingPost = getEditingPost(state);
+    const currentUserId = getCurrentUserId(state);
+    const channelId = editingPost?.post?.channel_id || getCurrentChannelId(state);
+    const teamId = getCurrentTeamId(state);
```
I am not 100% what the relation looks like ... could you elaborate on that a bit so I can get the change in? 🙇🏼

hmhealey replied:
I think we'd get the channel for the post and then get the team ID from that. If the channel is a DM/GM, then we could use the current team ID

hmhealey replied:
@michelengelen As I mentioned at the end of our meeting, it turns out that this prop actually isn't used by the component, so this line can just be removed

michelengelen replied:
Sorry, but I seem to not get it.

When I remove this line I have no params for the functions to follow that. 

Could you please explain what you mean? 
THX!!! 🙇🏼

hmhealey replied:
We don't actually need to pass the `teamId` in here at all since it's not used by the EditPost. This line can just be removed since it's not needed

michelengelen replied:
yeah, that's true, but I need the `teamId` for the `haveIChannelPermission` function

michelengelen commented on components/edit_post/index.ts:
```diff
@@ -0,0 +1,66 @@
+        editingPost,
+        channelId,
+        maxPostSize: parseInt(config.MaxPostSize || '0', 10) || Constants.DEFAULT_CHARACTER_LIMIT,
+        readOnlyChannel: !isCurrentUserSystemAdmin(state) && config.ExperimentalTownSquareIsReadOnly === 'true' && channel.name === Constants.DEFAULT_CHANNEL,
```
done

michelengelen commented on components/post_view/post_body/index.js:
```diff
@@ -16,10 +19,15 @@ function mapStateToProps(state, ownProps) {
         parentPost,
         parentPostUser,
         pluginPostTypes: state.plugins.postTypes,
+        isPostBeingEdited: isInlinePostEditingEnabled && getIsPostBeingEdited(state, ownProps.post.id),
```
good idea ... done the change

michelengelen commented on components/post_view/post/index.ts:
```diff
@@ -68,6 +76,9 @@ function makeMapStateToProps() {
             post,
+
+            // TODO@Michel: remove the call to `getIsInlinePostEditingEnabled` once inline post editing is enabled by default
+            isBeingEdited: getIsInlinePostEditingEnabled(state) && editingPost.postId === post.id,
```
done

michelengelen commented on components/post_view/post_reaction/post_reaction.tsx:
```diff
@@ -28,7 +28,7 @@ type Props = {
     postId: string;
     teamId: string;
-    getDotMenuRef: () => HTMLDivElement;
+    getDotMenuRef: () => HTMLDivElement | null;
```
It got pointed out as a problem when I removed the type-casting from post_info.tsx ... I reverted that change, although I am not 100% comfortable with that.

michelengelen commented on components/suggestion/suggestion_box/suggestion_box.jsx:
```diff
@@ -734,6 +734,14 @@ export default class SuggestionBox extends React.PureComponent {
         this.container = container;
     };
 
+    getListposition = (listPosition) => {
```
done

michelengelen commented on components/suggestion/suggestion_box/suggestion_box.jsx:
```diff
@@ -798,7 +806,7 @@ export default class SuggestionBox extends React.PureComponent {
                             open={this.state.focused || this.props.forceSuggestionsWhenBlur}
                             pretext={this.pretext}
-                            position={listPosition}
+                            position={this.getListposition(listPosition)}
```
done

michelengelen commented:
@hmhealey I left some of your comments unresolved since those might need some more info/discussion ... Let me know what you think of those if you find the time! 🙇🏼

hmhealey commented:
@michelengelen I think I've responded to everything above. Feel free to ping me if I missed anything.

jgilliam17 commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/106978

jgilliam17 commented:
@michelengelen I was able to test a cloud bot server using this commit as version. I ran the E2Es using the slash command, forgetting  this was behind a feature flag, so re-running and will check the report tomorrow.
Noticed few outstanding items while manually testing:
1. Editing a long post adds quite a bit of empty white space below post being edited when RHS is open
<img width="1179" alt="Screen Shot 2022-01-20 at 10 24 26 PM" src="https://user-images.githubusercontent.com/52937121/150460442-771a4f77-8adc-4d42-bfa2-5a9486a2e0e1.png">

2. Editing a post with autocomplete e.g. emoji causes autocomplete to show again after selection is made and user types additional characters

https://user-images.githubusercontent.com/52937121/150460676-f42d5cf2-13f9-4294-a98d-e5dfefc7f286.mov

3. System message is counted on thread footer. 
This is not reproducing on master and I am not entirely certain that this is related to your changes, noting as present.

<img width="1185" alt="Screen Shot 2022-01-20 at 10 37 30 PM" src="https://user-images.githubusercontent.com/52937121/150461495-374dba7e-319e-44c2-adba-b9724fc94156.png">

jgilliam17 commented:
/e2e-test MM_ENV="MM_FEATUREFLAGS_INLINEPOSTEDITING=true"

mattermod commented:
Triggering e2e testing with options:
```&map[MM_ENV:MM_FEATUREFLAGS_INLINEPOSTEDITING=true]```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/108293

jgilliam17 commented:
/update-branch

jgilliam17 commented:
Thanks @michelengelen  I re-ran the E2Es mostly so I can try using feature flag with the slash command
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/108293-5b75f74-onprem-ent-webapp-pr-8882-mattermost/mm-ee-test5b75f74/mochawesome.html) looks good, no PR related failures.
Looks and works great.
I can file these as follow up if needed so we can have this PR merged. Let me know.
Seeing 2 small issue
1. When selecting a post to edit, the inline edit modal jumps to the bottom of the screen.
This is happening on Desktop, Chrome, Safari and Edge, but not on Firefox.

https://user-images.githubusercontent.com/52937121/151626373-bbe3497a-8b47-4b18-9688-930f97d32efb.mov


2. When editing a long post, inline edit modal opens at the top, while cursor is at the end of the post  so some scrolling is necessary. This may be as design due to the length  of the post. Noting here as observed.

michelengelen commented:
@jgilliam17 I "fixed" that jumping behaviour ... to do that I had to change how the editbox renders a bit ... the good thing about this: I could add some animations to support the state change! 👍🏼

jgilliam17 commented:
Thanks @michelengelen Is the fix ready for re-test? Let me know. 🙂

michelengelen commented:
> Thanks @michelengelen Is the fix ready for re-test? Let me know. 🙂 

Yep @jgilliam17 ... go ahead!

matthewbirtch commented:
> @jgilliam17 I "fixed" that jumping behaviour ... to do that I had to change how the editbox renders a bit ... the good thing about this: I could add some animations to support the state change! 👍🏼

@michelengelen the new animation feels a bit awkward to me now that is slides in from left-to-right. Can we simplify that to just expand the area to fit the input box (like the [figma prototype](https://www.figma.com/proto/fVLSLjGls2d2LUvCqk2JMS/MM-29137-Message-Editing?page-id=0%3A1&node-id=524%3A17015&viewport=478%2C48%2C0.07&scaling=min-zoom&starting-point-node-id=524%3A17015&show-proto-sidebar=1))?

matthewbirtch commented:
@michelengelen this is definitely an improvement. However, it doesn't feel overly smooth - it feels a bit jittery to me, but I'm not quite sure what's causing that. I also think the animation could be a bit faster.

jgilliam17 commented:
Thanks @michelengelen 

1. When post is selected for edit on the RHS, same post also gets highlighted on the main view 

<img width="1341" alt="Screen Shot 2022-02-03 at 9 04 10 PM" src="https://user-images.githubusercontent.com/52937121/152461053-7129a1c0-6b6f-4340-aa95-a48e7ecf5dba.png">

2. CRT off, Consecutive post small styling issue, break in the line 
<img width="656" alt="Screen Shot 2022-02-03 at 9 21 05 PM" src="https://user-images.githubusercontent.com/52937121/152461938-d799fd13-3608-4fcc-8ebd-b1642ef70066.png">

3. Emoji picker and autocompletes are opening inside editbox
<img width="639" alt="Screen Shot 2022-02-03 at 9 30 54 PM" src="https://user-images.githubusercontent.com/52937121/152462742-52e5e0cd-0f72-4685-b4bb-c234285ff6b1.png">
<img width="643" alt="Screen Shot 2022-02-03 at 9 33 41 PM" src="https://user-images.githubusercontent.com/52937121/152462891-bf032482-0440-4301-bfeb-7133aa1df8c6.png">

matthewbirtch commented:
@michelengelen it's MUCH smoother and I'm good with this transition now. 

However, I discovered a new problem. When in an RHS thread, If I use the `UP` key, which is the shortcut to edit my last post, it seems to mess up the whole layout (see scree recording). If I edit the same post by clicking 'edit' from the post menu, I don't get the same problem though.

It's very strange, because it only seems to be happening with this one particular post in this thread
 
![2022-02-04 14 05 35](https://user-images.githubusercontent.com/2040554/152588587-7b1b66ab-1474-41f5-b711-116d80d2eddd.gif)

matthewbirtch commented:
@michelengelen had another look. A few things:

1. Although the transition does seem smoother, it still doesn't feel quite right to me. When I click edit on a long post (that has the 'show more' button), the container first seems to get shorter and then grow accordingly rather than expand. This makes it feel a bit jankey. (see GIF below)

2. In the RHS, when I edit a post that is closer to the top and then cancel (by pressing ESC), the content scrolls all the way to the bottom of the RHS (see GIF below)

<img src="https://user-images.githubusercontent.com/2040554/153073370-07addcc1-fe7f-438f-ad3f-1b4d25fe0df2.gif" width="500">

3. I also discovered another bug with some odd scrolling behavior and the RHS header and Channel Header shift under the global header

![2022-02-08 15 53 56](https://user-images.githubusercontent.com/2040554/153074145-ae9954eb-173c-47fb-a630-7aad43f7dfa6.gif)

michelengelen commented:
@matthewbirtch I might need to build my very own solution in that case ... the problems here are always due to the content being replaced, which is a behaviour that is not very common and there is no real solution to that ... I will think of something and come back later

hmhealey commented:
> 3. I also discovered another bug with some odd scrolling behavior and the RHS header and Channel Header shift under the global header

@michelengelen If you pull the latest master, this might be fixed. We ended up reverting some earlier changes to the root CSS that caused similar bugs elsewhere with the whole page scrolling being allowed to scroll

matthewbirtch reviewed (changes requested):
This is feeling SO GOOD now @michelengelen. Only two things I noticed that might need a little work:

1. When editing a root post in the RHS thread view, that same post gets highlighted in the center channel. I think that's a bit distracting. Is it possible to avoid that?
2. Also, in the RHS, when editing a post that is higher up in the thread, if I hit ESC to cancel the edit, it scrolls back down the the bottom of the thread upon canceling which was disorienting for me

michelengelen commented:
>     1. When editing a root post in the RHS thread view, that same post gets highlighted in the center channel. I think that's a bit distracting. Is it possible to avoid that?

I can fix that ... no problem.
 
>     2. Also, in the RHS, when editing a post that is higher up in the thread, if I hit ESC to cancel the edit, it scrolls back down the the bottom of the thread upon canceling which was disorienting for me

thats just because the textinput to add a comment to the thread is being focussed after the editing is complete. this is expected browser behaviour. It will be irrelevant once we add the PR #9482 to the app. in the meantime we could just not refocus the textinput, but thats kind of against the flow and we might need to re-add that once we merge the sticky reply PR.

matthewbirtch commented:
> thats just because the textinput to add a comment to the thread is being focussed after the editing is complete. this is expected browser behaviour. It will be irrelevant once we add the PR #9482 to the app. in the meantime we could just not refocus the textinput, but thats kind of against the flow and we might need to re-add that once we merge the sticky reply PR.

Ahhh, right. That makes sense. Thanks for the explanation @michelengelen

I'm going to approve then!

michelengelen commented:
@jgilliam17 this is now (finally) ready for QA review

jgilliam17 commented:
/e2e-test MM_ENV="MM_FEATUREFLAGS_INLINEPOSTEDITING=true"

mattermod commented:
Triggering e2e testing with options:
```&map[MM_ENV:MM_FEATUREFLAGS_INLINEPOSTEDITING=true]```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/150892

jgilliam17 commented:
Thanks @michelengelen 
Looks really good, this is the only issue still outstanding 
1. Emoji picker and autocompletes are opening inside editbox (see [comment](https://github.com/mattermost/mattermost-webapp/pull/8882#issuecomment-1029589235) )

michelengelen commented:
> Thanks @michelengelen Looks really good, this is the only issue still outstanding
> 
>     1. Emoji picker and autocompletes are opening inside editbox (see [comment](https://github.com/mattermost/mattermost-webapp/pull/8882#issuecomment-1029589235) )

@jgilliam17 fixed!

jgilliam17 commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/151227

jgilliam17 commented:
Thanks @michelengelen for a quick fix
1. When editing on pinned or saved RHS post, autocomplete is opening inside the edit box. Edit on main channel view and threads is working as expected.

<img width="450" alt="Screen Shot 2022-02-15 at 11 17 13 AM" src="https://user-images.githubusercontent.com/52937121/154103542-6397cc61-afdf-42ee-963c-8157d748fea0.png">

2. There is bit of editbox movement on rhs when mouse hovering. See video.


https://user-images.githubusercontent.com/52937121/154104610-62a67b68-3bcd-4a8c-812a-729706d4ca0f.mov


Few tests are failing [when feature flag is off](https://automation-dashboard.vercel.app/cycles/dd02599c-f095-47a3-a8cf-a358578c52d4). Can you please take a look?

michelengelen commented:
@jgilliam17 I fixed the issues you found and will look at fixing the e2e tests now

michelengelen commented:
@jgilliam17 I could only find 2 tests that were failing due to changes from this PR. One that failed because it could not find the element due to a missing class and the second one failed because the modal from the first failing test was still open! :D

I will now fix the pipeline and merge the current master. Could you restart e2e tests and review the issues you found once again please? 

Thanks!!! 🙇🏼

jgilliam17 commented:
/e2e-test MM_ENV="MM_FEATUREFLAGS_INLINEPOSTEDITING=true"

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/154309

michelengelen commented:
@jgilliam17 The issues/failing tests should now be fixed. I overdid it with my effort to improve render-performance and the updates did not get caught by the component, so the messages did stay the same although they should have been updated :D

jgilliam17 commented:
E2Es are running [here](https://automation-dashboard.vercel.app/cycles/3a5839b9-01a0-41d3-9a97-abbba3170627)

jgilliam17 commented:
Thanks @michelengelen LGTM
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/154870-7bb63ba-onprem-ent-webapp-pr-8882-mm-ee-test:7bb63ba/mochawesome.html) looks good, no PR related failures

jgilliam17 commented:
@michelengelen some check are still failing, not able to merge. Please take a look. Thanks

jgilliam17 commented:
/update-branch

michelengelen commented:
@amyblais @esethna this is merged now! 🎉 (finally)
